File: hackerrank/problem solving/data structures/1 - arrays/5 - sparse arrays/sparse_arrays.rb
# https://www.hackerrank.com/challenges/sparse-arrays/problem?isFullScreen=true

def main
  n = gets.to_i
  strings = read_lines(n)
  strings_to_counts = strings.tally

  n = gets.to_i
  queries = read_lines(n)

  puts counts_from(strings_to_counts, queries)
end

def read_lines(n)
  n.times.map { gets.strip }
end

# n: length of queries
# max_query_string_length: 20
# T: O(n * max_query_string_length) = O(n * 20) = O(n)
# S: O(n) extra space
def counts_from(freq_map, queries)
  queries.map { |q| freq_map[q] || 0 }
end

main if __FILE__ == $0

File: hackerrank/problem solving/data structures/1 - arrays/5 - sparse arrays/sparse_arrays.rs
// https://www.hackerrank.com/challenges/sparse-arrays/problem?isFullScreen=true

use counter::Counter;

fn main() {
    let mut n: usize = text_io::read!();
    let strings: Vec<String> = read_lines(n);
    let strings_to_counts = strings.iter().collect::<Counter<_>>();

    n = text_io::read!();
    let queries: Vec<String> = read_lines(n);

    counts_from(strings_to_counts, &queries)
        .iter()
        .for_each(|x| println!("{}", x))
}

fn read_lines(n: usize) -> Vec<String> {
    (0..n).map(|_| text_io::read!()).collect()
}

// n: length of queries
// maxQueryStringLength: 20
// T: O(n * maxQueryStringLength) = O(n * 20) = O(n)
// S: O(n) extra space
fn counts_from(freq_map: Counter<&String>, queries: &[String]) -> Vec<usize> {
    queries.iter().map(|q| freq_map[q]).collect()
}
